sdk/search/azure-search/src/main/java/com/azure/search/SearchIndexClient.java

```diff
@@ -0,0 +1,354 @@
+    }
+
+    /**
+     * Uploads a collection of documents to the target index
+     *
+     * @param documents collection of documents to upload to the target Index.
+     * @param context additional context that is passed through the Http pipeline during the service call
+     * @return response containing the document index result.
+     */
+    public Response<IndexDocumentsResult> uploadDocumentsWithResponse(Iterable<?> documents, Context context) {
+        return asyncClient.uploadDocumentsWithResponse(documents, context).block();
+    }
+
+    /**
+     * Merges a collection of documents with existing documents in the target index.
+     *
+     * @param documents collection of documents to be merged
+     * @return document index result
+     */
+    public IndexDocumentsResult mergeDocuments(Iterable<?> documents) {
+        return this.mergeDocumentsWithResponse(documents, Context.NONE).getValue();
+    }
+
+    /**
+     * Merges a collection of documents with existing documents in the target index.
+     *
+     * @param documents collection of documents to be merged
+     * @param context additional context that is passed through the Http pipeline during the service call
+     * @return response containing the document index result.
+     */
+    public Response<IndexDocumentsResult> mergeDocumentsWithResponse(Iterable<?> documents, Context context) {
+        return asyncClient.mergeDocumentsWithResponse(documents, context).block();
+    }
+
+    /**
+     * This action behaves like merge if a document with the given key already exists in the index.
+     * If the document does not exist, it behaves like upload with a new document.
+     *
+     * @param documents collection of documents to be merged, if exists, otherwise uploaded
+     * @return document index result
+     */
+    public IndexDocumentsResult mergeOrUploadDocuments(Iterable<?> documents) {
+        return this.mergeOrUploadDocumentsWithResponse(documents, Context.NONE).getValue();
+    }
+
+    /**
+     * This action behaves like merge if a document with the given key already exists in the index.
+     * If the document does not exist, it behaves like upload with a new document.
+     *
+     * @param documents collection of documents to be merged, if exists, otherwise uploaded
+     * @param context additional context that is passed through the Http pipeline during the service call
+     * @return response containing a document index result
+     */
+    public Response<IndexDocumentsResult> mergeOrUploadDocumentsWithResponse(Iterable<?> documents, Context context) {
+        return asyncClient.mergeOrUploadDocumentsWithResponse(documents, context).block();
+    }
+
+    /**
+     * Deletes a collection of documents from the target index
+     *
+     * @param documents collection of documents to delete from the target Index.
+     * @return document index result.
+     */
+    public IndexDocumentsResult deleteDocuments(Iterable<?> documents) {
```

Comment:
@navalev There is another way to delete documents in the .NET SDK, by providing the name of the key field and a list of document keys:

https://github.com/Azure/azure-sdk-for-net/blob/8f8b4f5452c0521bfb71636730c947000e44b5f8/sdk/search/Microsoft.Azure.Search.Data/src/Customizations/Documents/Models/IndexBatch.Customization.cs#L22

I don't recall whether we considered having a similar overload on the Java side, or if we decided not to during API review. Do you remember how this was decided, or if we just missed it?